bundles/io/org.eclipse.smarthome.io.voice/src/main/java/org/eclipse/smarthome/io/voice/text/AbstractRuleBasedInterpreter.java

```diff
@@ -106,23 +107,89 @@ public String interpret(Locale locale, String text) throws InterpretationExcepti
     }
 
     private void invalidate() {
-        identifierTokens = null;
+        allItemTokens = null;
+        itemTokens = null;
         languageRules = null;
     }
 
     /**
      * All the tokens (name parts) of the names of all the items in the {@link ItemRegistry}.
      *
+     * @param locale The locale that is to be used for preparing the tokens.
      * @return the identifier tokens
      */
-    HashSet<String> getIdentifierTokens() {
-        if (identifierTokens == null) {
-            identifierTokens = new HashSet<String>();
+    HashSet<String> getAllItemTokens(Locale locale) {
+        if (allItemTokens == null) {
+            allItemTokens = new HashMap<Locale, HashSet<String>>();
+        }
+        HashSet<String> localeTokens = allItemTokens.get(locale);
+        if (localeTokens == null) {
+            allItemTokens.put(locale, localeTokens = new HashSet<String>());
             for (Item item : itemRegistry.getAll()) {
-                identifierTokens.addAll(splitName(item.getLabel(), true));
+                localeTokens.addAll(tokenize(locale, item.getLabel()));
+            }
+        }
+        return localeTokens;
+    }
+
+    /**
+     * Retrieves the list of identifier token sets per item currently contained in the {@link ItemRegistry}.
+     *
+     * @param locale The locale that is to be used for preparing the tokens.
+     * @return the list of identifier token sets per item
+     */
+    HashMap<Item, ArrayList<HashSet<String>>> getItemTokens(Locale locale) {
+        if (itemTokens == null) {
+            itemTokens = new HashMap<Locale, HashMap<Item, ArrayList<HashSet<String>>>>();
```

Comment:
This lazy initialization is following a dirty pattern. Whenever things in the item DB are added/removed/changed, some members are nulled:
```java
private void invalidate() {
    allItemTokens = null;
    itemTokens = null;
    languageRules = null;
}
```
`invalidate` could also instantiate `itemTokens`, but this would hurt the symmetry of the pattern.